### PR TITLE
Update documentation and license text.

### PR DESCRIPTION
**Pull request by ashman-p**
Added doc for Stateful Signature, XMSS, LMS and update licenses per file in LMS directories.

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the list of algorithms available -- either adding, removing, or renaming? Does this PR otherwise change an API? (If so, PRs in fully supported downstream projects dependent on these, i.e., [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) and [OQS-OpenSSH](https://github.com/open-quantum-safe/openssh) will also need to be ready for review and merge by the time this is merged.)

### CODE REVIEW

**Review by baentsch — COMMENTED** [review]
Not having claimed SecLevel/NISTLevel may cause problems downstream. But not sure, so no blocker.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (dstebila) [inline docs/algorithms/sig_stfl/sig_stfl.md:1]: Good idea to add this file!  Can you add a little bit of markdown formatting to it?

File: src/sig_stfl/lms/external/hash.c [hunk 2]
```diff
@@ -32,7 +33,8 @@ void hss_hash_ctx(void *result, int hash_type, union hash_context *ctx,
           const void *message, size_t message_len) {
 #if ALLOW_VERBOSE
     if (hss_verbose) {
-        int i; for (i=0; i< message_len; i++) printf( " %02x%s", ((unsigned char*)message)[i], (i%16 == 15) ? "\n" : "" );
+        int i; for (i=0; i< message_len; i++) printf( " %02x%s", ((unsigned char*)message)[i], (i%16 == 15) ? "
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hash.c:36]: Did something go wrong with the `\n` symbols here?

File: src/sig_stfl/lms/external/hash.c [hunk 3]
```diff
@@ -44,7 +46,8 @@ void hss_hash_ctx(void *result, int hash_type, union hash_context *ctx,
 #if ALLOW_VERBOSE
         if (hss_verbose) {
             printf( " ->" );
-            int i; for (i=0; i<32; i++) printf( " %02x", ((unsigned char *)result)[i] ); printf( "\n" );
+            int i; for (i=0; i<32; i++) printf( " %02x", ((unsigned char *)result)[i] ); printf( "
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hash.c:49]: Did something go wrong with the `\n` symbols here?

File: src/sig_stfl/lms/external/hash.c [hunk 4]
```diff
@@ -95,7 +98,8 @@ void hss_finalize_hash_context(int h, union hash_context *ctx, void *buffer) {
     if (hss_verbose) {
         printf( " -->" );
         int i; for (i=0; i<32; i++) printf( " %02x", ((unsigned char*)buffer)[i] );
-        printf( "\n" );
+        printf( "
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hash.c:101]: Did something go wrong with the `\n` symbols here?

File: src/sig_stfl/lms/external/hss_alloc.c [hunk 5]
```diff
@@ -417,7 +418,8 @@ signed long initial_mem_target = mem_target; /* DEBUG HACK */
         return 0;
     }
 #if 0
-printf( "Allocation = %ld\n", initial_mem_target - mem_target + best_mem ); /* DEBUG HACK */
+printf( "Allocation = %ld
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hss_alloc.c:421]: Did something go wrong with the `\n` symbol here?

File: src/sig_stfl/lms/external/hss_sign_inc.h [hunk 6]
```diff
@@ -22,7 +23,8 @@
  *    hss_sign_update( &ctx, message_part_2, len_2 );
  *    hss_sign_update( &ctx, message_part_3, len_3 );
  *    success = hss_sign_finalize( &ctx, working_key, signature );
- *    if (success) printf( "We generated the signature\n" );
+ *    if (success) printf( "We generated the signature
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hss_sign_inc.h:26]: Did something go wrong with the `\n` symbol here?

File: src/sig_stfl/lms/external/hss_verify_inc.h [hunk 7]
```diff
@@ -20,7 +21,8 @@
  *    hss_validate_update( &ctx, message_part_2, len_2 );
  *    hss_validate_update( &ctx, message_part_3, len_3 );
  *    success = hss_validate_finalize( &ctx, signature );
- *    if (success) printf( "The signature validated\n" );
+ *    if (success) printf( "The signature validated
```
Comment (dstebila) [inline src/sig_stfl/lms/external/hss_verify_inc.h:24]: Did something go wrong with the `\n` symbol here?

File: src/sig_stfl/lms/external/hash.c (hunk 2 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hash.c:36]: Yep, will fix.

File: src/sig_stfl/lms/external/hash.c (hunk 3 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hash.c:49]: Yep, will fix.

File: src/sig_stfl/lms/external/hash.c (hunk 4 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hash.c:101]: Yep, will fix.

File: src/sig_stfl/lms/external/hss_alloc.c (hunk 5 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hss_alloc.c:421]: Yep, will fix.

File: src/sig_stfl/lms/external/hss_sign_inc.h (hunk 6 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hss_sign_inc.h:26]: Yep, will fix.

File: src/sig_stfl/lms/external/hss_verify_inc.h (hunk 7 above)
Comment (ashman-p) [reply src/sig_stfl/lms/external/hss_verify_inc.h:24]: Yep, will fix.

File: docs/algorithms/sig_stfl/lms.yml [hunk 8]
```diff
@@ -0,0 +1,214 @@
+name: LMS
+type: stateful signature
+principal-submitters:
+- Scott Fluhrer
+auxiliary-submitters:
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/lms.yml:5]: Scott is the only author on the code.

File: docs/algorithms/sig_stfl/lms.yml [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+crypto-assumption: stateful hash-based signatures
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/lms.yml:7]: Updated with "second preimage resistance of SHA-256"

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+Stateful Hash Based Signature
+
+The security of hash based signatures (HBS) is based on the underlaying hash functions on which they are built.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:4]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+NIST recommendation is that they are suitable for near term use to mitigate against attacks mounted by quantum computers.
+While not a general purpose solution, they are useful means to authenticate boot or firmware images.
+
+This package provides full support for a variety of variants for XMSS and LMS. 
+Key generation, signature generation, and signature verification. 
+Security of HBS also depends on the management of the state of the secret key. Secret keys can only used once to generate a signature.
+Multiple signing with same key can reveal that key to an attacker.
+Because of this, NIST recommends that key and signature aeration be done in hardware security modules.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:12]: Fixed (generation)

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+Having said that, this library is fully functional for research purposes. Storage and management of the secret keys are left update applications using this feature.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:13]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+Secret key storage is easily done when by supplying a callback function to the library.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:14]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+
+Key State Management:
+Application writers have to supply callback functions to store and update secret keys.
+After a sign operation the secret key index is advanced and stored. This ensures one-time use of the key.
+Signing operations will fail without this callback set because the private key cannot be advance (to prevent reuse).
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:20]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+Stateful keys can generate a finite number signatures. A counter tracks the limit when the key is created and is decremented after each signature is generated.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:22]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+When the counter is down to 0, signature generations fails. Applications can query the remaining count via an API.
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:23]: Fixed.

File: docs/algorithms/sig_stfl/sig_stfl.md [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+Multithreaded application will need to set lock and unlock callbacks for the library to use. This is needed to ensure that only one thread can access a given secret key
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:25]: Fixed.

File: docs/algorithms/sig_stfl/xmss.yml [hunk 18]
```diff
@@ -0,0 +1,187 @@
+name: XMSS
+type: stateful signature
+principal-submitters:
+- Joost Rijneveld
+- A. Huelsing
+- David Cooper
+- Bas Westerbaan
+auxiliary-submitters:
+
+crypto-assumption: stateful hash-based signatures
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/xmss.yml]: Updated as suggested.

File: docs/algorithms/sig_stfl/xmss.yml [hunk 19]
```diff
@@ -0,0 +1,187 @@
+name: XMSS
+type: stateful signature
+principal-submitters:
+- Joost Rijneveld
+- A. Huelsing
+- David Cooper
+- Bas Westerbaan
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/xmss.yml:7]: Code authors listed.

File: docs/algorithms/sig_stfl/sig_stfl.md (hunk 1 above)
Comment (ashman-p) [reply docs/algorithms/sig_stfl/sig_stfl.md:1]: Thanks. Done.

File: docs/algorithms/sig_stfl/lms.yml [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+website: https://www.rfc-editor.org/info/rfc8554
+nist-round: 
+spec-version: 
+spdx-license-identifier: 
+primary-upstream:
+  source: 
+  spdx-license-identifier: MIT
```
[the remark this answers is not in the document]
Comment (ashman-p) [reply docs/algorithms/sig_stfl/lms.yml:16]: Added source.

**Review by SWilson4 — APPROVED** [review]

**Pull request merged**